### PR TITLE
refactor: make `{Map, MutMap, RedBlackTree}.{query, queryWith}` effect polymorphic

### LINKED ISSUE #3972: add effects to `Map`/`MutMap`/`RedBlackTree` `.` `query`/`queryWith`
example:
```
pub def query(p: k -> Comparison, m: Map[k, v]): List[(k, v)] =
        let Map(t) = m;
        RedBlackTree.query(p, (k, v) -> (k, v), t)
```
`p` should have an effect.

### PR DESCRIPTION
**Pull request by ninaandrup**
Closes #3972

### CODE REVIEW

File: main/src/library/RedBlackTree.flix [hunk 1]
```diff
@@ -562,9 +562,9 @@ namespace RedBlackTree {
     ///
     /// That is, the result is a list of all pairs `f(k, v)` where `p(k)` returns `Equal`.
     ///
-    pub def query(p: k -> Comparison, f: (k, v) -> a & ef, t: RedBlackTree[k, v]): List[a] & ef = {
+    pub def query(p: k -> Comparison & ef, f: (k, v) -> a & ef, t: RedBlackTree[k, v]): List[a] & ef = {
         let buffer = new MutList(Static);
-        let g = k -> v -> MutList.push!(f(k, v), buffer);
+        let g = k -> v -> MutList.push!(f(k, v), buffer) as & ef;
```
Comment (magnus-madsen) [inline main/src/library/RedBlackTree.flix:567]: This could use a region.

**Comment by magnus-madsen** [comment]
I see a lot of casts. Are these really necessary?

**Comment by ninaandrup** [comment]
> I see a lot of casts. Are these really necessary?

I think I know how to solve it

**Comment by ninaandrup** [comment]
> 

Nope, they weren't necessary :-)

**Pull request merged**